compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java

```diff
@@ -197,6 +235,137 @@ private void performCodeGen(boolean shrink) {
     }
 
+    private void registerUnusedBIRNodes() {
+        // Reversed the for loop because used BIRNode analysis should start from the root module.
+        // Root module is usually found last in the topologicallySortedModuleList.
+        for (int i = pkgResolution.topologicallySortedModuleList().size() - 1; i >= 0; i--) {
+            ModuleContext moduleContext = pkgResolution.topologicallySortedModuleList().get(i);
+
+            // Default module is analyzed first to find its immediate dependencies.
+            // Its immediate dependent modules are marked as "used" and they are optimized after that.
+            // This process happens till all "used" modules are exhausted.
+            if (shouldOptimize(moduleContext) && (isRootModule(moduleContext) || moduleContext.isUsed())) {
+                this.usedBIRNodeAnalyzer.analyze(moduleContext.bLangPackage());
+                updateNativeDependencyMap(moduleContext);
+            }
+        }
+
+        CodeGenOptimizationReportEmitter.emitBirOptimizationDuration();
+
+        if (this.packageContext.project().buildOptions().verbose()) {
+            CodeGenOptimizationReportEmitter.emitCodegenOptimizationReport(
+                    this.usedBIRNodeAnalyzer.pkgWiseInvocationData, getOptimizationReportPath(),
+                    packageContext.project().kind());
+        }
+    }
+
+    private Path getOptimizationReportPath() {
+        return this.packageContext.project().sourceRoot;
+    }
+
+    protected Set<PackageID> getOptimizedPackageIDs() {
+        return this.usedBIRNodeAnalyzer.pkgWiseInvocationData.keySet();
+    }
+
+    private void markTestDependenciesForDuplicateBIRGen() {
+        for (int i = pkgResolution.topologicallySortedModuleList().size() - 1; i >= 0; i--) {
+            ModuleContext moduleContext = pkgResolution.topologicallySortedModuleList().get(i);
+            BLangPackage bLangPackage = moduleContext.bLangPackage();
+
+            if (isRootModule(moduleContext) && bLangPackage.hasTestablePackage()) {
+                Set<BPackageSymbol> buildPkgDependencies = new HashSet<>();
+                Set<BPackageSymbol> testablePkgDependencies = new HashSet<>();
+                collectDependencies(bLangPackage.symbol, buildPkgDependencies);
+                collectDependencies(bLangPackage.getTestablePkg().symbol, testablePkgDependencies);
+
+                buildPkgDependencies.stream()
+                        .filter(testablePkgDependencies::contains).filter(this::isNotWhiteListedPkg)
+                        .forEach(pkgSymbol -> {
+                            pkgSymbol.shouldGenerateDuplicateBIR = true;
+                            // Have to use a hashmap because the pkgIds get mutated later
+                            JvmCodeGenUtil.duplicatePkgsMap.put(
+                                    pkgSymbol.pkgID.orgName + pkgSymbol.pkgID.getNameComps().toString(),
+                                    pkgSymbol.pkgID);
+                        });
```

Comment:
Extract this into separate method to reduce depth in this method

Reply:
Fixed in 93679cfa7b33eb81a194f1b170448cafc2b12c9b.